File: controllers/expense_opponent_controller.py
from repository.expense_opponent_repository import ExpenseOpponentRepository
from utils.logger_utils import log_error, log_info
from fastapi import HTTPException
from datetime import datetime as dt

class ExpenseOpponentController:
    """
    Controller for handling expense opponent related operations.
    """

    @staticmethod
    async def create_expense_opponent(expense_opponent_data: dict, userID: int):
        """
        Create an expense opponent in the database.
        """
        log_info(f"Creating expense opponent with data: {expense_opponent_data}")
        try:
            expense_opponent_data["createdAt"] = dt.now()
            expense_opponent_data["createdBy"] = userID
            expense_opponent_data["isDelete"] = False
            
            result = await ExpenseOpponentRepository.create(expense_opponent_data)
            if "error" in result:
                log_error(f"Error creating expense opponent: {result['error']}")
                raise HTTPException(status_code=result["status"], detail=result["error"])
            
            return {"message": "Expense opponent created successfully", "opponent_id": result.get("opponent_id")}
        except HTTPException:
            raise
        except Exception as e:
            log_error(f"Unexpected error creating expense opponent: {str(e)}")
            raise HTTPException(status_code=500, detail="Internal server error")
        
    @staticmethod
    async def get_expense_opponents(page: int, pageSize: int, sortBy: str | None, sortByDirection: str | None, keyword: str | None):
        """
        Retrieve a list of expense opponents from the database.
        """
        if page < 1:
            return {"error": "Page number must be greater than 0", "status": 400}
        
        log_info(f"Retrieving expense opponents with page={page}, pageSize={pageSize}, sortBy={sortBy}, sortByDirection={sortByDirection}, keyword={keyword}")
        
        try:
            result = await ExpenseOpponentRepository.get_all(page, pageSize, sortBy, sortByDirection, keyword)
            if "error" in result:
                log_error(f"Error retrieving expense opponents: {result['error']}")
                raise HTTPException(status_code=result["status"], detail=result["error"])
            
            return result
        except HTTPException:
            raise
        except Exception as e:
            log_error(f"Unexpected error retrieving expense opponents: {str(e)}")
            raise HTTPException(status_code=500, detail="Internal server error")

    @staticmethod
    async def get_expense_opponent_by_id(opponent_id: int):
        """
        Get an expense opponent by ID.
        """
        log_info(f"Retrieving expense opponent with ID: {opponent_id}")
        
        try:
            result = await ExpenseOpponentRepository.get_by_id(opponent_id)
            if "error" in result:
                log_error(f"Error retrieving expense opponent: {result['error']}")
                raise HTTPException(status_code=result["status"], detail=result["error"])
            
            return result
        except HTTPException:
            raise
        except Exception as e:
            log_error(f"Unexpected error retrieving expense opponent: {str(e)}")
            raise HTTPException(status_code=500, detail="Internal server error")

    @staticmethod
    async def update_expense_opponent(opponent_id: int, opponent_data: dict, userID: int):
        """
        Update an expense opponent.
        """
        log_info(f"Updating expense opponent {opponent_id} with data: {opponent_data}")
        
        try:
            opponent_data["updatedBy"] = userID
            opponent_data["updatedAt"] = dt.now()
            
            result = await ExpenseOpponentRepository.update(opponent_id, opponent_data)
            if "error" in result:
                log_error(f"Error updating expense opponent: {result['error']}")
                raise HTTPException(status_code=result["status"], detail=result["error"])
            
            return result
        except HTTPException:
            raise
        except Exception as e:
            log_error(f"Unexpected error updating expense opponent: {str(e)}")
            raise HTTPException(status_code=500, detail="Internal server error")

    @staticmethod
    async def delete_expense_opponent(opponent_id: int, userID: int):
        """
        Delete an expense opponent.
        """
        log_info(f"Deleting expense opponent with ID: {opponent_id}")
        
        try:
            result = await ExpenseOpponentRepository.delete(opponent_id, userID)
            if "error" in result:
                log_error(f"Error deleting expense opponent: {result['error']}")
                raise HTTPException(status_code=result["status"], detail=result["error"])
            
            return result
        except HTTPException:
            raise
        except Exception as e:
            log_error(f"Unexpected error deleting expense opponent: {str(e)}")
            raise HTTPException(status_code=500, detail="Internal server error")

    @staticmethod
    async def search_expense_opponents(keyword: str, limit: int = 10):
        """
        Search expense opponents by name.
        """
        log_info(f"Searching expense opponents with keyword: {keyword}")
        
        try:
            result = await ExpenseOpponentRepository.search_by_name(keyword, limit)
            if "error" in result:
                log_error(f"Error searching expense opponents: {result['error']}")
                raise HTTPException(status_code=result["status"], detail=result["error"])
            
            return result
        except HTTPException:
            raise
        except Exception as e:
            log_error(f"Unexpected error searching expense opponents: {str(e)}")
            raise HTTPException(status_code=500, detail="Internal server error")

File: src/controllers/expenseOpponent.controller.ts
import { ExpenseOpponentRepository } from "../repository/expenseOpponent"
import { UserRepository } from "../repository/user"
import { logError, logInfo } from "../utils/logger"
import { paginationMeta } from "../utils/pagination"
import { ExpenseOpponentModel } from "../models/expenseOpponent.model"
import type { IExpenseOpponentCreate, IExpenseOpponentOutput, IExpenseOpponentUpdate, IExpenseOpponentView } from "../interfaces/expenseOpponent.interfaces"
import type { IUserView } from "../interfaces/user.interfaces"

async function buildUserView(userId: number | null | undefined): Promise<IUserView> {
  if (!userId) return { id: 0, name: "Unknown" }
  const user = await UserRepository.getById(userId)
  return { id: user?.id ?? 0, name: user?.name ?? "Unknown" }
}

export class ExpenseOpponentController {
  static async create(data: IExpenseOpponentCreate, userId: number): Promise<{ kind: "success"; data: IExpenseOpponentOutput } | { error: string; status: number }> {
    try {
      const payload = ExpenseOpponentModel.createPayload(data)
      const record = await ExpenseOpponentRepository.create({ ...payload, createdBy: userId, createdAt: new Date(), isDelete: false })
      const createdBy = await buildUserView(userId)
      return { kind: "success", data: ExpenseOpponentModel.toOutput(record, createdBy, null) }
    } catch (e) {
      logError(`Error creating expense opponent: ${e}`)
      return { error: "Failed to create expense opponent", status: 500 }
    }
  }

  static async getAll(page: number, pageSize: number, keyword?: string, sortBy = "name", sortDir = "asc"): Promise<{ data: IExpenseOpponentView[]; meta: any } | { error: string; status: number }> {
    if (page < 1) return { error: "Page must be >= 1", status: 400 }
    try {
      const { data, total } = await ExpenseOpponentRepository.getAll(page, pageSize, keyword, sortBy, sortDir)
      return { data: data.map(ExpenseOpponentModel.toView), meta: paginationMeta(total, page, pageSize) }
    } catch (e) {
      logError(`Error fetching expense opponents: ${e}`)
      return { error: "Failed to fetch expense opponents", status: 500 }
    }
  }

  static async getById(id: number): Promise<{ kind: "success"; data: IExpenseOpponentOutput } | { error: string; status: number }> {
    try {
      const record = await ExpenseOpponentRepository.getById(id)
      if (!record) return { error: "Expense opponent not found", status: 404 }
      const [createdBy, updatedBy] = await Promise.all([
        buildUserView(record.createdBy),
        buildUserView(record.updatedBy),
      ])
      return { kind: "success", data: ExpenseOpponentModel.toOutput(record, createdBy, updatedBy) }
    } catch (e) {
      logError(`Error fetching expense opponent ${id}: ${e}`)
      return { error: "Failed to fetch expense opponent", status: 500 }
    }
  }

  static async update(id: number, data: IExpenseOpponentUpdate, userId: number): Promise<{ kind: "success"; data: IExpenseOpponentOutput } | { error: string; status: number }> {
    try {
      const existing = await ExpenseOpponentRepository.getById(id)
      if (!existing) return { error: "Expense opponent not found", status: 404 }
      const payload = ExpenseOpponentModel.updatePayload(data)
      const updated = await ExpenseOpponentRepository.update(id, { ...payload, updatedBy: userId, updatedAt: new Date() })
      const [createdBy, updatedBy] = await Promise.all([
        buildUserView(existing.createdBy),
        buildUserView(userId),
      ])
      return { kind: "success", data: ExpenseOpponentModel.toOutput(updated, createdBy, updatedBy) }
    } catch (e) {
      logError(`Error updating expense opponent ${id}: ${e}`)
      return { error: "Failed to update expense opponent", status: 500 }
    }
  }

  static async delete(id: number, userId: number): Promise<{ message: string } | { error: string; status: number }> {
    try {
      const existing = await ExpenseOpponentRepository.getById(id)
      if (!existing) return { error: "Expense opponent not found", status: 404 }
      await ExpenseOpponentRepository.softDelete(id, userId)
      return { message: "Expense opponent deleted successfully" }
    } catch (e) {
      logError(`Error deleting expense opponent ${id}: ${e}`)
      return { error: "Failed to delete expense opponent", status: 500 }
    }
  }

  static async search(keyword: string, limit = 20): Promise<IExpenseOpponentView[] | { error: string; status: number }> {
    try {
      const results = await ExpenseOpponentRepository.search(keyword, limit)
      return results.map(ExpenseOpponentModel.toView)
    } catch (e) {
      logError(`Error searching expense opponents: ${e}`)
      return { error: "Failed to search expense opponents", status: 500 }
    }
  }
}
